dynomite/src/lib.rs

```diff
@@ -274,39 +308,94 @@ pub trait Attribute: Sized {
 }
 
+impl Attribute for AttributeValue {
+    fn into_attr(self: Self) -> AttributeValue {
+        self
+    }
+    fn from_attr(value: AttributeValue) -> Result<Self, AttributeError> {
+        Ok(value)
+    }
+}
+
 /// A type capable of being produced from
 /// a set of string keys and `AttributeValues`
 pub trait FromAttributes: Sized {
+    /// Shortcut for `FromAttributes::from_mut_attrs(&mut attrs)`.
+    /// You should generally implement only that method.
+    fn from_attrs(mut attrs: Attributes) -> Result<Self, AttributeError> {
+        Self::from_mut_attrs(&mut attrs)
+    }
+
     /// Returns an instance of of a type resolved at runtime from a collection
-    /// of a `String` keys and `AttributeValues`. If
-    /// a instance can not be resolved and `AttributeError` will be returned.
-    fn from_attrs(attrs: Attributes) -> Result<Self, AttributeError>;
+    /// of a `String` keys and `AttributeValues`.
+    /// If an instance can not be resolved and `AttributeError` will be returned.
+    /// The implementations of this method should remove the relevant key-value
+    /// pairs from the map to consume them. This is needed to support
+    /// `#[dynomite(flatten)]` without creating temporary hash maps.
+    fn from_mut_attrs(attrs: &mut Attributes) -> Result<Self, AttributeError>;
 }
 
 /// Coerces a homogenious HashMap of attribute values into a homogeneous Map of types
 /// that implement Attribute
 #[allow(clippy::implicit_hasher)]
 impl<A: Attribute> FromAttributes for HashMap<String, A> {
-    fn from_attrs(attrs: Attributes) -> Result<Self, AttributeError> {
+    fn from_mut_attrs(attrs: &mut Attributes) -> Result<Self, AttributeError> {
         attrs
-            .into_iter()
-            .try_fold(HashMap::new(), |mut result, (k, v)| {
-                result.insert(k, A::from_attr(v)?);
-                Ok(result)
-            })
+            .drain()
+            .map(|(k, v)| Ok((k, A::from_attr(v)?)))
+            .collect()
     }
 }
 
 /// Coerces a homogenious Map of attribute values into a homogeneous BTreeMap of types
 /// that implement Attribute
 impl<A: Attribute> FromAttributes for BTreeMap<String, A> {
-    fn from_attrs(attrs: Attributes) -> Result<Self, AttributeError> {
+    fn from_mut_attrs(attrs: &mut Attributes) -> Result<Self, AttributeError> {
         attrs
-            .into_iter()
-            .try_fold(BTreeMap::new(), |mut result, (k, v)| {
-                result.insert(k, A::from_attr(v)?);
-                Ok(result)
-            })
+            .drain()
+            .map(|(k, v)| Ok((k, A::from_attr(v)?)))
+            .collect()
+    }
+}
+
+/// You should implement this trait instead of `From<T> for Attributes`
+/// for your type to support flattening, `#[dynomite(Attributes/Item)]` will
+/// generate both the implementation of this trait and `From<T>`
+/// (there is no blanket impl for `From<T>` here due to orphan rules)
```

Comment:
Would it be possible to use derive Attributes instead of introducting a new trait?